File: src/main/java/com/algorithm/chaptertwo/sort/first/SelectionSort.java
package com.algorithm.chaptertwo.sort.first;

import com.algorithm.chaptertwo.sort.util.SortUtils;

/**
 * @author clx 2018/7/2.
 */
public class SelectionSort {

    /**
     * selection sort
     *
     * @param array
     */
    public static void sort(Comparable<Integer>[] array) {
        if (SortUtils.checkArrayValid(array)) {
            return;
        }
        int length = array.length;
        int minIdx;
        for (int idx = 0; idx < length; idx++) {
            minIdx = idx;
            for (int jdx = idx + 1; jdx < length; jdx++) {
                if (SortUtils.less(array[jdx], array[minIdx])) {
                    minIdx = jdx;
                }
            }
            SortUtils.exchange(array, idx, minIdx);
        }
    }
}

File: algorithm-go/src/grokkingalgorithms/secondchapter/selection_sort.go
package secondchapter

import "grokkingalgorithms/common"

// selection sort
func SelectionSort(array []int) {
	if array == nil || len(array) == 0 {
		return
	}

	minIdx := 0
	for idx := range array {
		minIdx = idx
		for jdx := idx + 1; jdx < len(array); jdx++ {
			if common.Less(array[jdx], array[minIdx]) {
				minIdx = jdx
			}
		}
		common.Exchange(array, idx, minIdx)
	}
}
